numpy/core/src/umath/simd.inc.src

```diff
@@ -1270,6 +1352,137 @@ static NPY_INLINE NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ @vtype@
  */
 
 
+/*
+ * Vectorized approximate sine/cosine algorithms: The following code is a
+ * vectorized version of the algorithm presented here:
+ * https://stackoverflow.com/questions/30463616/payne-hanek-algorithm-implementation-in-c/30465751#30465751
+ * (1) Load data in ZMM/YMM registers and generate mask for elements that are
+ * within range [-71476.0625f, 71476.0625f] for cosine and [-117435.992f,
+ * 117435.992f] for sine.
+ * (2) For elements within range, perform range reduction using Cody-Waite's
+ * method: x* = x - y*PI/2, where y = rint(x*2/PI). x* \in [-PI/4, PI/4].
+ * (3) Map cos(x) to (+/-)sine or (+/-)cosine of x* based on the quadrant k =
+ * int(y).
+ * (4) For elements outside that range, Cody-Waite reduction peforms poorly
+ * leading to catastrophic cancellation. We compute cosine by calling glibc in
+ * a scalar fashion.
+ * (5) Vectorized implementation has a max ULP of 1.49 and performs at least
+ * 5-7x faster than scalar implementations when magnitude of all elements in
+ * the array < 71476.0625f (117435.992f for sine). Worst case performance is
+ * when all the elements are large leading to about 1-2% reduction in
+ * performance.
+ */
+
+#if defined HAVE_ATTRIBUTE_TARGET_@ISA@_WITH_INTRINSICS
+static NPY_GCC_OPT_3 NPY_GCC_TARGET_@ISA@ void
+@ISA@_sincos_FLOAT(npy_float * op, npy_float * ip, const npy_int array_size,
+                                                   char* operation)
+{
+    const npy_int num_lanes = @BYTES@/sizeof(npy_float);
+    npy_int compute_cos = 1;
+    npy_float large_number = 71476.0625f;
+    if (*operation == 's') {
```

Comment:
an enum would be better to select the operation

Reply:
@r-devulap ping

Reply:
oh, sorry this almost slipped my mind. I will add an enum and fix this.

Reply:
Is there a reason why enum is better than char comparison? (just for my understanding)

Reply:
Because the enum documents intent better, in this case by making it clear what the complete set of allowing values is

Reply:
thanks, added a commit to use enum instead.